docs/design/redundant-suppressions.md

```diff
@@ -0,0 +1,93 @@
+
+### Example:
+Let us again consider the example of the trimmer-compatible code with a redundant warning suppression. 
+
+```csharp
+    [UnconditionalSuppressMessage("trim", "IL2075", Justification = "It's OK to print only the properties which were actually used.")]
+    void PrintProperties<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicProperties)] T>(T instance)
+    {
+        foreach (var p in typeof(T).GetProperties())
+        {
+            PrintPropertyValue(p, instance);
+        }
+    }
+
+```
+
+In order to detect the warning suppression not tied to any trimmer-incompatible pattern we should run the `dotnet publish` command and pass the `--check-suppressions` option.
+```shell
+  dotnet publish -r win-x64 -p:PublishTrimmed=True p:_ExtraTrimmerArgs="--check-suppressions"
+```
+
+The warning should be reported in the output of the command.
+
+```
+Trim analysis warning IL2021: Program.PrintProperties<T>(T): Unused UnconditionalSuppressMessageAttribute suppressing the IL2075 warning found. Consider removing the unused warning suppression.
+```
+
+## Other solutions
+
+The proposed solution operates by extending the functionality of the linker tool. On one hand, this allows for reusing the existing components leading to a simple implementation, on the other hand this may lead to potential problems. The linker sees only a part of code which is actually used, that means that the solution would not be able to identify the warning suppressions on the code which is trimmed away. Also, the linker may visit parts of the code which were not authored by the developer (e.g. used libraries) and look for the redundant suppressions there. This may cause the output warnings to be noisy and not useful to the developer. Moreover, the dependencies identified by the linker may be different depending on the environment the tool is run in. Hence, the proposed solution may report different redundant suppressions on different environments.
+
+Alternatively, we could make the analyzer do the unused warning suppressions detection. The analyzer operates on a single assembly level but it has a full view of the processed assembly, as opposed to the linker which only sees the code which is actually used within the assembly. Making the solution part of the analyzer, would allow us then to detect the unused warning suppression on assembly level with higher precision. Also, an advantage of such solution would be a shorter feedback loop; we would learn about the redundant suppressions way before we run the publish command. The drawback of this approach is the added complexity. We would not be able to reuse the existing components. What is more, the analyzer has a different view of the code than the linker. We may not be able to identify the same set of trimmer-incompatible patterns using the analyzer as we do using the linker.
```

Comment:
I would be worried that the analyzer can recommend removing a suppression just because doesn't have the full picture. Non-actionable code and code that either way is going to be removed seems less worrying to me. 
On the linker side of things, I would be worried if it's implementing the linker tooling is going to recommend removing a suppression because of something we annotated using a feature that the analyzer knows best (like backtracking).

Reply:
Since we are already talking about things that are blurry in the document here:
What's the scope of this feature? my guess is that it should be restricted to a certain amount of warnings and attributes. For example, if it's implemented in the linker we wouldn't be able to detect a redundant UnconditionalSuppressMessage originated by RequiresDynamicCode or RequiresAssemblyFiles, or what is worse it will be flagged always as redundant just because Linker doesn't analyze it. 
Also, is this feature going to be supported by NativeAOT?

Reply:
Added a note that only suppressions produced by the linker will be checked.